rust/src/instruction/mod.rs

```diff
@@ -0,0 +1,217 @@
+use pyo3::{
+    exceptions::{PyRuntimeError, PyValueError},
+    prelude::*,
+    types::PyBytes,
+};
+
+use std::{collections::HashMap, str::FromStr};
+
+use quil_rs::quil::Quil;
+
+mod declaration;
+mod gate;
+mod quilt;
+
+pub use declaration::Declare;
+pub use gate::*;
+pub use quilt::*;
+
+use crate::impl_eq;
+
+pub fn init_module(_py: Python<'_>) -> PyResult<Bound<'_, PyModule>> {
+    let m = PyModule::new_bound(_py, "instruction")?;
+    m.add_class::<Instruction>()?;
+    m.add_class::<DefCalibration>()?;
+    m.add_class::<DefMeasureCalibration>()?;
+    m.add_class::<GateModifier>()?;
+    Ok(m)
+}
+
+#[pyclass(subclass)]
+#[derive(Clone, Debug, PartialEq)]
+pub struct Instruction {
+    inner: quil_rs::instruction::Instruction,
+}
+impl_eq!(Instruction);
+
+impl From<Instruction> for quil_rs::instruction::Instruction {
+    fn from(instruction: Instruction) -> Self {
+        instruction.inner
+    }
+}
+
+impl From<quil_rs::instruction::Instruction> for Instruction {
+    fn from(instruction: quil_rs::instruction::Instruction) -> Self {
+        Self { inner: instruction }
+    }
+}
+
+#[pymethods]
+impl Instruction {
+    fn out(&self) -> PyResult<String> {
+        self.inner
+            .to_quil()
+            .map_err(|e| PyValueError::new_err(e.to_string()))
+    }
+
+    pub fn __str__(&self) -> String {
+        self.inner.to_quil_or_debug()
+    }
+
+    pub fn __repr__(&self) -> String {
+        self.inner.to_quil_or_debug()
+    }
+
+    pub fn __deepcopy__(self_: PyRef<'_, Self>, _memo: &pyo3::types::PyDict) -> Self {
+        let mut instruction = Self {
+            inner: self_.inner.clone(),
+        };
+
+        let mut placeholders: HashMap<
+            quil_rs::instruction::QubitPlaceholder,
+            quil_rs::instruction::QubitPlaceholder,
+        > = HashMap::new();
+
+        for qubit in instruction.inner.get_qubits_mut() {
+            match qubit {
+                quil_rs::instruction::Qubit::Fixed(_)
+                | quil_rs::instruction::Qubit::Variable(_) => *qubit = qubit.clone(),
+                quil_rs::instruction::Qubit::Placeholder(placeholder) => {
+                    *qubit = quil_rs::instruction::Qubit::Placeholder(
+                        placeholders.entry(placeholder.clone()).or_default().clone(),
```

Comment:
Will add a comment, but basically, `QubitPlaceholders` are `Arc<()>`, so just cloning them would create a copy to the same pointer. So if the copy is resolved, the placeholders in the original would also be resolved. For a deepcopy, that's an interaction we want to avoid.